dtfs-central-api/src/v1/controllers/utilisation-report-service/post-add-payment.controller/helpers.test.ts

```diff
@@ -0,0 +1,206 @@
+      currency: paymentCurrency,
+      amountReceived: 100,
+      dateReceived: new Date(),
+      paymentReference: 'A payment reference',
+    };
+
+    const utilisationReport = UtilisationReportEntityMockBuilder.forStatus('RECONCILIATION_IN_PROGRESS').withId(reportId).build();
+    const utilisationReportStateMachine = UtilisationReportStateMachine.forReport(utilisationReport);
+
+    const feeRecordsInPaymentCurrency = feeRecordIds.map((feeRecordId) =>
+      FeeRecordEntityMockBuilder.forReport(utilisationReport).withId(feeRecordId).withPaymentCurrency(paymentCurrency).build(),
+    );
+
+    const mockQueryRunner = getMockQueryRunner();
+    const feeRecordFindBySpy = jest.spyOn(FeeRecordRepo, 'findBy');
+    const createQueryRunnerSpy = jest.spyOn(SqlDbDataSource, 'createQueryRunner');
+
+    const utilisationReportStateMachineConstructorSpy = jest.spyOn(UtilisationReportStateMachine, 'forReportId');
+    const handleEventSpy = jest.spyOn(utilisationReportStateMachine, 'handleEvent');
+
+    beforeEach(() => {
+      feeRecordFindBySpy.mockResolvedValue(feeRecordsInPaymentCurrency);
+      createQueryRunnerSpy.mockReturnValue(mockQueryRunner);
+      utilisationReportStateMachineConstructorSpy.mockResolvedValue(utilisationReportStateMachine);
+      handleEventSpy.mockResolvedValue(utilisationReport);
+    });
+
+    afterEach(() => {
+      jest.resetAllMocks();
+    });
+
+    it("calls the 'UtilisationReportStateMachine.forReportId' method with the supplied report id", async () => {
+      // Act
+      await addPaymentToUtilisationReport(reportId, feeRecordIds, tfmUser, newPaymentDetails);
+
+      // Assert
+      expect(utilisationReportStateMachineConstructorSpy).toHaveBeenCalledWith(reportId);
+    });
+
+    it("calls the 'FeeRecordRepo.findBy' method with the supplied list of fee record ids", async () => {
+      // Act
+      await addPaymentToUtilisationReport(reportId, feeRecordIds, tfmUser, newPaymentDetails);
+
+      // Assert
+      expect(feeRecordFindBySpy).toHaveBeenCalledWith({ id: In(feeRecordIds) });
+    });
+
+    it("throws the 'NotFoundError' if no fee records are found", async () => {
+      // Arrange
+      feeRecordFindBySpy.mockResolvedValue([]);
+
+      // Act / Assert
+      await expect(addPaymentToUtilisationReport(reportId, feeRecordIds, tfmUser, newPaymentDetails)).rejects.toThrow(NotFoundError);
+    });
+
+    it("throws the 'InvalidPayloadError' if the payment currency does not match the fee record payment currencies", async () => {
+      // Arrange
+      const feeRecordsWithGBPPaymentCurrency = feeRecordIds.map((feeRecordId) =>
+        FeeRecordEntityMockBuilder.forReport(utilisationReport).withId(feeRecordId).withPaymentCurrency('GBP').build(),
+      );
+      feeRecordFindBySpy.mockResolvedValue(feeRecordsWithGBPPaymentCurrency);
+
+      const newPaymentDetailsWithEURCurrency: NewPaymentDetails = {
+        ...newPaymentDetails,
+        currency: 'EUR',
+      };
```

Comment:
Can we add a duplicate of this test in which all but one of the fee records matches the payment currency?